File: Token/offchain/src/rpc.rs
use std::fmt;

use crate::config::ProtocolConfig;
use crate::indexer::{
    deposit_topic, node_updated_topic, protocol_config_updated_topic, ref_bound_topic, RawLog,
};
use crate::state::Node;
use ethers_core::types::Address;
use serde::{Deserialize, Serialize};

const OWNER_SELECTOR: &str = "0x8da5cb5b";

#[derive(Debug, Clone)]
pub struct BscRpcClient {
    client: reqwest::Client,
    rpc_url: String,
    token_address: Address,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PairReserves {
    pub pair: String,
    pub token_reserve: u128,
    pub bnb_reserve: u128,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ChainProtocolConfig {
    pub operator: String,
    pub buy_enabled: bool,
    pub config: ProtocolConfig,
}

#[derive(Debug)]
pub enum RpcError {
    Http(reqwest::Error),
    Rpc(String),
    InvalidAddress,
    InvalidHex,
    MissingResult,
}

impl fmt::Display for RpcError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{self:?}")
    }
}

impl std::error::Error for RpcError {}

impl From<reqwest::Error> for RpcError {
    fn from(error: reqwest::Error) -> Self {
        Self::Http(error)
    }
}

impl BscRpcClient {
    pub fn new(rpc_url: impl Into<String>, token_address: &str) -> Result<Self, RpcError> {
        let token_address = parse_address(token_address)?;
        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(15))
            .connect_timeout(std::time::Duration::from_secs(10))
            .build()
            .map_err(RpcError::Http)?;
        Ok(Self {
            client,
            rpc_url: rpc_url.into(),
            token_address,
        })
    }

    pub fn token_address(&self) -> Address {
        self.token_address
    }

    pub async fn owner(&self) -> Result<Address, RpcError> {
        let output = self.eth_call(OWNER_SELECTOR).await?;
        parse_address_word(&output)
    }

    pub async fn vault(&self) -> Result<Address, RpcError> {
        let data = function_selector("vault()");
        let output = self.eth_call(&data).await?;
        parse_address_word(&output)
    }

    pub async fn pair(&self) -> Result<Address, RpcError> {
        let data = function_selector("pair()");
        let output = self.eth_call(&data).await?;
        parse_address_word(&output)
    }

    pub async fn protocol_config(&self) -> Result<ChainProtocolConfig, RpcError> {
        let output = self
            .eth_call(&function_selector("getProtocolConfig()"))
            .await?;
        let mut team_reward_bps = [0u16; 10];
        for (index, rate) in team_reward_bps.iter_mut().enumerate() {
            *rate = parse_u16_word(&output, 15 + index)?;
        }
        let config = ProtocolConfig {
            min_deposit: parse_u128_word(&output, 3)?,
            max_deposit: parse_u128_word(&output, 4)?,
            lp_build_bps: parse_u16_word(&output, 6)?,
            node_bps: parse_u16_word(&output, 7)?,
            builder_buy_bps: parse_u16_word(&output, 8)?,
            vault_bps: parse_u16_word(&output, 9)?,
            direct_pool_bps: parse_u16_word(&output, 10)?,
            direct_reward_bps: parse_u16_word(&output, 11)?,
            daily_static_bps: parse_u16_word(&output, 12)?,
            settlement_periods_per_day: parse_u8_word(&output, 13)?,
            exit_multiple_bps: parse_u32_word(&output, 14)?,
            team_reward_bps,
            deflation_enabled: parse_bool_word(&output, 25)?,
            deflation_hourly_bps: parse_u16_word(&output, 26)?,
            deflation_daily_cap_bps: parse_u16_word(&output, 27)?,
            buyback_enabled: parse_bool_word(&output, 28)?,
            buyback_per_minute: parse_u128_word(&output, 29)?,
            buy_tax_bps: parse_u16_word(&output, 1)?,
            buy_tax_builder_bps: parse_u16_word(&output, 30)?,
            buy_tax_vault_bps: parse_u16_word(&output, 31)?,
            sell_tax_bps: parse_u16_word(&output, 2)?,
            sell_tax_builder_bps: parse_u16_word(&output, 32)?,
            sell_tax_owner_bps: parse_u16_word(&output, 33)?,
            sell_tax_vault_bps: parse_u16_word(&output, 34)?,
        };
        config
            .validate()
            .map_err(|error| RpcError::Rpc(format!("invalid protocol config: {error}")))?;
        Ok(ChainProtocolConfig {
            operator: format_address(parse_address_word_at(&output, 0)?),
            buy_enabled: parse_bool_word(&output, 5)?,
            config,
        })
    }

    pub async fn nodes(&self) -> Result<Vec<Node>, RpcError> {
        let count_output = self.eth_call(&function_selector("nodeCount()")).await?;
        let count = usize::try_from(parse_u128_word(&count_output, 0)?)
            .map_err(|_| RpcError::InvalidHex)?;
        let mut nodes = Vec::with_capacity(count);
        for index in 0..count {
            let data = format!(
                "{}{}",
                function_selector("nodeAt(uint256)"),
                u256_word(index as u128)
            );
            let output = self.eth_call(&data).await?;
            let address = format_address(parse_address_word_at(&output, 0)?);
            let weight = parse_u32_word(&output, 1)?;
            if weight != 0 {
                nodes.push(Node { address, weight });
            }
        }
        Ok(nodes)
    }

    pub async fn pair_reserves(&self) -> Result<Option<PairReserves>, RpcError> {
        let pair = self.pair().await?;
        if pair == Address::zero() {
            return Ok(None);
        }
        let token0 = parse_address_word(
            &self
                .eth_call_to(pair, &function_selector("token0()"))
                .await?,
        )?;
        let reserves = self
            .eth_call_to(pair, &function_selector("getReserves()"))
            .await?;
        let reserve0 = parse_u128_word(&reserves, 0)?;
        let reserve1 = parse_u128_word(&reserves, 1)?;
        let (token_reserve, bnb_reserve) = if token0 == self.token_address {
            (reserve0, reserve1)
        } else {
            (reserve1, reserve0)
        };
        Ok(Some(PairReserves {
            pair: format_address(pair),
            token_reserve,
            bnb_reserve,
        }))
    }

    pub async fn block_number(&self) -> Result<u64, RpcError> {
        let response = self
            .client
            .post(&self.rpc_url)
            .json(&JsonRpcRequest {
                jsonrpc: "2.0",
                id: 1,
                method: "eth_blockNumber",
                params: Vec::<serde_json::Value>::new(),
            })
            .send()
            .await?
            .error_for_status()?
            .json::<JsonRpcResponse<String>>()
            .await?;
        if let Some(error) = response.error {
            return Err(RpcError::Rpc(error.message));
        }
        let result = response.result.ok_or(RpcError::MissingResult)?;
        parse_hex_u64(&result)
    }

    pub async fn block_hash(&self, block_number: u64) -> Result<String, RpcError> {
        let response = self
            .client
            .post(&self.rpc_url)
            .json(&JsonRpcRequest {
                jsonrpc: "2.0",
                id: 1,
                method: "eth_getBlockByNumber",
                params: vec![
                    serde_json::Value::String(hex_quantity(block_number)),
                    serde_json::Value::Bool(false),
                ],
            })
            .send()
            .await?
            .error_for_status()?
            .json::<JsonRpcResponse<Option<JsonRpcBlock>>>()
            .await?;
        if let Some(error) = response.error {
            return Err(RpcError::Rpc(error.message));
        }
        response
            .result
            .flatten()
            .map(|block| block.hash.to_ascii_lowercase())
            .ok_or(RpcError::MissingResult)
    }

    pub async fn protocol_logs(
        &self,
        from_block: u64,
        to_block: u64,
    ) -> Result<Vec<RawLog>, RpcError> {
        let filter = serde_json::json!({
            "address": format_address(self.token_address),
            "fromBlock": hex_quantity(from_block),
            "toBlock": hex_quantity(to_block),
            "topics": [[
                ref_bound_topic(),
                deposit_topic(),
                protocol_config_updated_topic(),
                node_updated_topic(),
            ]],
        });
        let response = self
            .client
            .post(&self.rpc_url)
            .json(&JsonRpcRequest {
                jsonrpc: "2.0",
                id: 1,
                method: "eth_getLogs",
                params: vec![filter],
            })
            .send()
            .await?
            .error_for_status()?
            .json::<JsonRpcResponse<Vec<JsonRpcLog>>>()
            .await?;
        if let Some(error) = response.error {
            return Err(RpcError::Rpc(error.message));
        }
        response
            .result
            .ok_or(RpcError::MissingResult)?
            .into_iter()
            .filter(|log| !log.removed.unwrap_or(false))
            .map(raw_log_from_rpc)
            .collect()
    }

    async fn eth_call(&self, data: &str) -> Result<String, RpcError> {
        self.eth_call_to(self.token_address, data).await
    }

    async fn eth_call_to(&self, target: Address, data: &str) -> Result<String, RpcError> {
        let call = serde_json::json!({
            "to": format_address(target),
            "data": data,
        });
        let response = self
            .client
            .post(&self.rpc_url)
            .json(&JsonRpcRequest {
                jsonrpc: "2.0",
                id: 1,
                method: "eth_call",
                params: vec![call, serde_json::Value::String("latest".to_owned())],
            })
            .send()
            .await?
            .error_for_status()?
            .json::<JsonRpcResponse<String>>()
            .await?;
        if let Some(error) = response.error {
            return Err(RpcError::Rpc(error.message));
        }
        response.result.ok_or(RpcError::MissingResult)
    }
}

#[derive(Debug, Serialize)]
struct JsonRpcRequest<P> {
    jsonrpc: &'static str,
    id: u64,
    method: &'static str,
    params: P,
}

#[derive(Debug, Deserialize)]
struct JsonRpcResponse<T> {
    result: Option<T>,
    error: Option<JsonRpcError>,
}

#[derive(Debug, Deserialize)]
struct JsonRpcError {
    message: String,
}

#[derive(Debug, Deserialize)]
struct JsonRpcBlock {
    hash: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct JsonRpcLog {
    block_number: String,
    block_hash: String,
    transaction_hash: String,
    log_index: String,
    topics: Vec<String>,
    data: String,
    removed: Option<bool>,
}

fn raw_log_from_rpc(log: JsonRpcLog) -> Result<RawLog, RpcError> {
    let log_index = parse_hex_u64(&log.log_index)?;
    Ok(RawLog {
        block_number: parse_hex_u64(&log.block_number)?,
        block_hash: log.block_hash.to_ascii_lowercase(),
        tx_hash: log.transaction_hash.to_ascii_lowercase(),
        log_index: u32::try_from(log_index).map_err(|_| RpcError::InvalidHex)?,
        topics: log
            .topics
            .into_iter()
            .map(|topic| topic.to_ascii_lowercase())
            .collect(),
        data: log.data,
    })
}

fn parse_hex_u64(value: &str) -> Result<u64, RpcError> {
    u64::from_str_radix(value.trim_start_matches("0x"), 16).map_err(|_| RpcError::InvalidHex)
}

fn hex_quantity(value: u64) -> String {
    format!("0x{value:x}")
}

fn function_selector(signature: &str) -> String {
    let hash = ethers_core::utils::keccak256(signature.as_bytes());
    format!("0x{}", hex_encode(&hash[..4]))
}

fn hex_encode(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut output = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        output.push(HEX[(byte >> 4) as usize] as char);
        output.push(HEX[(byte & 0x0f) as usize] as char);
    }
    output
}

pub fn parse_address(value: &str) -> Result<Address, RpcError> {
    let trimmed = value
        .trim()
        .strip_prefix("0x")
        .ok_or(RpcError::InvalidAddress)?;
    if trimmed.len() != 40 || !trimmed.chars().all(|char| char.is_ascii_hexdigit()) {
        return Err(RpcError::InvalidAddress);
    }
    let mut bytes = [0u8; 20];
    for index in 0..20 {
        bytes[index] = u8::from_str_radix(&trimmed[index * 2..index * 2 + 2], 16)
            .map_err(|_| RpcError::InvalidHex)?;
    }
    Ok(Address::from(bytes))
}

pub fn parse_address_word(value: &str) -> Result<Address, RpcError> {
    let trimmed = value
        .trim()
        .strip_prefix("0x")
        .ok_or(RpcError::InvalidHex)?;
    if trimmed.len() != 64 || !trimmed.chars().all(|char| char.is_ascii_hexdigit()) {
        return Err(RpcError::InvalidHex);
    }
    parse_address(&format!("0x{}", &trimmed[24..]))
}

fn parse_address_word_at(value: &str, index: usize) -> Result<Address, RpcError> {
    parse_address(&format!("0x{}", &word_at(value, index)?[24..]))
}

fn parse_u128_word(value: &str, index: usize) -> Result<u128, RpcError> {
    let word = word_at(value, index)?;
    if word[..32].chars().any(|char| char != '0') {
        return Err(RpcError::InvalidHex);
    }
    u128::from_str_radix(&word[32..], 16).map_err(|_| RpcError::InvalidHex)
}

fn parse_u32_word(value: &str, index: usize) -> Result<u32, RpcError> {
    let value = parse_u128_word(value, index)?;
    u32::try_from(value).map_err(|_| RpcError::InvalidHex)
}

fn parse_u16_word(value: &str, index: usize) -> Result<u16, RpcError> {
    let value = parse_u128_word(value, index)?;
    u16::try_from(value).map_err(|_| RpcError::InvalidHex)
}

fn parse_u8_word(value: &str, index: usize) -> Result<u8, RpcError> {
    let value = parse_u128_word(value, index)?;
    u8::try_from(value).map_err(|_| RpcError::InvalidHex)
}

fn parse_bool_word(value: &str, index: usize) -> Result<bool, RpcError> {
    match parse_u128_word(value, index)? {
        0 => Ok(false),
        1 => Ok(true),
        _ => Err(RpcError::InvalidHex),
    }
}

fn word_at(value: &str, index: usize) -> Result<&str, RpcError> {
    let trimmed = value
        .trim()
        .strip_prefix("0x")
        .ok_or(RpcError::InvalidHex)?;
    let start = index.checked_mul(64).ok_or(RpcError::InvalidHex)?;
    let end = start.checked_add(64).ok_or(RpcError::InvalidHex)?;
    let word = trimmed.get(start..end).ok_or(RpcError::InvalidHex)?;
    if !word.chars().all(|char| char.is_ascii_hexdigit()) {
        return Err(RpcError::InvalidHex);
    }
    Ok(word)
}

fn u256_word(value: u128) -> String {
    format!("{value:064x}")
}

pub fn format_address(address: Address) -> String {
    format!("{address:#x}")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_address_from_abi_word() {
        let word = "0x0000000000000000000000001111111111111111111111111111111111111111";
        assert_eq!(
            format_address(parse_address_word(word).unwrap()),
            "0x1111111111111111111111111111111111111111"
        );
    }

    #[test]
    fn rejects_bad_addresses() {
        assert_eq!(
            parse_address("0x123").unwrap_err().to_string(),
            "InvalidAddress"
        );
        assert_eq!(
            parse_address_word("0x123").unwrap_err().to_string(),
            "InvalidHex"
        );
    }

    #[test]
    fn parses_rpc_log_quantities() {
        assert_eq!(parse_hex_u64("0x2a").unwrap(), 42);
        assert_eq!(hex_quantity(42), "0x2a");
        assert_eq!(function_selector("owner()"), OWNER_SELECTOR);
        let output = format!("0x{:064x}{:064x}", 10u128, 20u128);
        assert_eq!(parse_u128_word(&output, 0).unwrap(), 10);
        assert_eq!(parse_u128_word(&output, 1).unwrap(), 20);
    }
}

File: Token/worker/src/rpc.ts
/**
 * BSC JSON-RPC read client. Port of Token/offchain/src/rpc.rs.
 * Uses fetch() to the operator RPC URL. ABI word parsing is manual (matching rpc.rs word_at/parse_* helpers)
 * to preserve exact decoding semantics, including the getProtocolConfig() word index map.
 */
import { keccak256, toHex } from "viem";
import { defaultProtocolConfig, validateConfig, type ProtocolConfig } from "./config";
import { ALL_TOPICS, type RawLog } from "./indexer";

export const OWNER_SELECTOR = "0x8da5cb5b"; // owner()
export const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";

export interface PairReserves {
  pair: string;
  tokenReserve: bigint;
  bnbReserve: bigint;
}

export interface ChainProtocolConfig {
  operator: string;
  buyEnabled: boolean;
  config: ProtocolConfig;
}

export class RpcError extends Error {}

function functionSelector(signature: string): string {
  return keccak256(toHex(signature)).slice(0, 10); // 0x + 8 hex = 4 bytes
}

function hexQuantity(value: bigint): string {
  return `0x${value.toString(16)}`;
}

function u256Word(value: bigint): string {
  return value.toString(16).padStart(64, "0");
}

function parseHexBig(value: string): bigint {
  return BigInt(value);
}

function normalizeAddress(value: string): string {
  const t = value.trim();
  if (!t.startsWith("0x")) throw new RpcError("InvalidAddress");
  const hex = t.slice(2);
  if (hex.length !== 40 || !/^[0-9a-fA-F]{40}$/.test(hex)) throw new RpcError("InvalidAddress");
  return `0x${hex.toLowerCase()}`;
}

function wordAt(value: string, index: number): string {
  const t = value.trim();
  if (!t.startsWith("0x")) throw new RpcError("InvalidHex");
  const hex = t.slice(2);
  const start = index * 64;
  const word = hex.slice(start, start + 64);
  if (word.length !== 64 || !/^[0-9a-fA-F]{64}$/.test(word)) throw new RpcError("InvalidHex");
  return word;
}

function parseU128Word(value: string, index: number): bigint {
  const word = wordAt(value, index);
  if (/[^0]/.test(word.slice(0, 32))) throw new RpcError("InvalidHex");
  return BigInt(`0x${word.slice(32)}`);
}

function parseNumWord(value: string, index: number, max: bigint): number {
  const v = parseU128Word(value, index);
  if (v > max) throw new RpcError("InvalidHex");
  return Number(v);
}

function parseBoolWord(value: string, index: number): boolean {
  const v = parseU128Word(value, index);
  if (v === 0n) return false;
  if (v === 1n) return true;
  throw new RpcError("InvalidHex");
}

function parseAddressWord(value: string): string {
  const t = value.trim();
  if (!t.startsWith("0x")) throw new RpcError("InvalidHex");
  const hex = t.slice(2);
  if (hex.length !== 64 || !/^[0-9a-fA-F]{64}$/.test(hex)) throw new RpcError("InvalidHex");
  return normalizeAddress(`0x${hex.slice(24)}`);
}

function parseAddressWordAt(value: string, index: number): string {
  return normalizeAddress(`0x${wordAt(value, index).slice(24)}`);
}

interface RpcLogJson {
  blockNumber: string;
  blockHash: string;
  transactionHash: string;
  logIndex: string;
  topics: string[];
  data: string;
  removed?: boolean;
}

export class BscRpcClient {
  private tokenAddress: string;
  private vaultCache: string | null = null;
  private pairCache: string | null = null;

  constructor(
    private rpcUrl: string,
    tokenAddress: string,
  ) {
    this.tokenAddress = normalizeAddress(tokenAddress);
  }

  getTokenAddress(): string {
    return this.tokenAddress;
  }

  private async rpc<T>(method: string, params: unknown[]): Promise<T> {
    const res = await fetch(this.rpcUrl, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ jsonrpc: "2.0", id: 1, method, params }),
    });
    if (!res.ok) throw new RpcError(`http ${res.status}`);
    const body = (await res.json()) as { result?: T; error?: { message: string } };
    if (body.error) throw new RpcError(body.error.message);
    if (body.result === undefined || body.result === null) throw new RpcError("MissingResult");
    return body.result;
  }

  private async ethCall(data: string): Promise<string> {
    return this.ethCallTo(this.tokenAddress, data);
  }

  private async ethCallTo(target: string, data: string): Promise<string> {
    return this.rpc<string>("eth_call", [{ to: normalizeAddress(target), data }, "latest"]);
  }

  async owner(): Promise<string> {
    return parseAddressWord(await this.ethCall(OWNER_SELECTOR));
  }

  async vault(): Promise<string> {
    if (this.vaultCache) return this.vaultCache;
    const vault = parseAddressWord(await this.ethCall(functionSelector("vault()")));
    this.vaultCache = vault;
    return vault;
  }

  async pair(): Promise<string> {
    if (this.pairCache) return this.pairCache;
    const pair = parseAddressWord(await this.ethCall(functionSelector("pair()")));
    if (pair !== ZERO_ADDRESS) this.pairCache = pair;
    return pair;
  }

  /** getProtocolConfig() — word index map is identical to rpc.rs:135-160. */
  async protocolConfig(): Promise<ChainProtocolConfig> {
    const out = await this.ethCall(functionSelector("getProtocolConfig()"));
    const U16 = 0xffffn;
    const U8 = 0xffn;
    const U32 = 0xffffffffn;

    const teamRewardBps: number[] = [];
    for (let i = 0; i < 10; i++) teamRewardBps.push(parseNumWord(out, 15 + i, U16));

    const config: ProtocolConfig = {
      ...defaultProtocolConfig(),
      minDeposit: parseU128Word(out, 3),
      maxDeposit: parseU128Word(out, 4),
      lpBuildBps: parseNumWord(out, 6, U16),
      nodeBps: parseNumWord(out, 7, U16),
      builderBuyBps: parseNumWord(out, 8, U16),
      vaultBps: parseNumWord(out, 9, U16),
      directPoolBps: parseNumWord(out, 10, U16),
      directRewardBps: parseNumWord(out, 11, U16),
      dailyStaticBps: parseNumWord(out, 12, U16),
      settlementPeriodsPerDay: parseNumWord(out, 13, U8),
      exitMultipleBps: parseNumWord(out, 14, U32),
      teamRewardBps,
      deflationEnabled: parseBoolWord(out, 25),
      deflationHourlyBps: parseNumWord(out, 26, U16),
      deflationDailyCapBps: parseNumWord(out, 27, U16),
      buybackEnabled: parseBoolWord(out, 28),
      buybackPerMinute: parseU128Word(out, 29),
      buyTaxBps: parseNumWord(out, 1, U16),
      buyTaxBuilderBps: parseNumWord(out, 30, U16),
      buyTaxVaultBps: parseNumWord(out, 31, U16),
      sellTaxBps: parseNumWord(out, 2, U16),
      sellTaxBuilderBps: parseNumWord(out, 32, U16),
      sellTaxOwnerBps: parseNumWord(out, 33, U16),
      sellTaxVaultBps: parseNumWord(out, 34, U16),
      bindCost: parseU128Word(out, 35),
    };
    try {
      validateConfig(config);
    } catch (e) {
      throw new RpcError(`invalid protocol config: ${(e as Error).message}`);
    }
    return {
      operator: parseAddressWordAt(out, 0),
      buyEnabled: parseBoolWord(out, 5),
      config,
    };
  }

  async nodes(): Promise<{ address: string; weight: number }[]> {
    const countOut = await this.ethCall(functionSelector("nodeCount()"));
    const count = Number(parseU128Word(countOut, 0));
    const nodes: { address: string; weight: number }[] = [];
    for (let i = 0; i < count; i++) {
      const data = functionSelector("nodeAt(uint256)") + u256Word(BigInt(i));
      const out = await this.ethCall(data);
      const address = parseAddressWordAt(out, 0);
      const weight = parseNumWord(out, 1, 0xffffffffn);
      if (weight !== 0) nodes.push({ address, weight });
    }
    return nodes;
  }

  async pairReserves(): Promise<PairReserves | null> {
    const pair = await this.pair();
    if (pair === ZERO_ADDRESS) return null;
    const token0 = parseAddressWord(await this.ethCallTo(pair, functionSelector("token0()")));
    const reserves = await this.ethCallTo(pair, functionSelector("getReserves()"));
    const reserve0 = parseU128Word(reserves, 0);
    const reserve1 = parseU128Word(reserves, 1);
    const [tokenReserve, bnbReserve] =
      token0 === this.tokenAddress ? [reserve0, reserve1] : [reserve1, reserve0];
    return { pair, tokenReserve, bnbReserve };
  }

  async nativeBalance(address: string): Promise<bigint> {
    const result = await this.rpc<string>("eth_getBalance", [normalizeAddress(address), "latest"]);
    return parseHexBig(result);
  }

  async tokenBalance(address: string): Promise<bigint> {
    const data = functionSelector("balanceOf(address)") + u256Word(BigInt(normalizeAddress(address)));
    const result = await this.ethCall(data);
    return parseU128Word(result, 0);
  }

  async blockNumber(): Promise<bigint> {
    return parseHexBig(await this.rpc<string>("eth_blockNumber", []));
  }

  async blockHash(blockNumber: bigint): Promise<string> {
    const block = await this.rpc<{ hash: string } | null>("eth_getBlockByNumber", [
      hexQuantity(blockNumber),
      false,
    ]);
    if (!block) throw new RpcError("MissingResult");
    return block.hash.toLowerCase();
  }

  async protocolLogs(fromBlock: bigint, toBlock: bigint): Promise<RawLog[]> {
    const filter = {
      address: this.tokenAddress,
      fromBlock: hexQuantity(fromBlock),
      toBlock: hexQuantity(toBlock),
      topics: [ALL_TOPICS],
    };
    const logs = await this.rpc<RpcLogJson[]>("eth_getLogs", [filter]);
    return logs
      .filter((log) => !log.removed)
      .map((log) => ({
        blockNumber: parseHexBig(log.blockNumber),
        blockHash: log.blockHash.toLowerCase(),
        txHash: log.transactionHash.toLowerCase(),
        logIndex: Number(parseHexBig(log.logIndex)),
        topics: log.topics.map((t) => t.toLowerCase()),
        data: log.data,
      }));
  }
}
